service/frontend/domainArchivalConfigStateMachine.go

```diff
@@ -21,216 +21,234 @@
+	// archivalEvent represents a change request to archival config state
+	// the only restriction placed on events is that defaultBucket is not empty
+	// setting requestedBucket to empty string means user is not attempting to set bucket
+	// status can be nil, enabled, or disabled (nil indicates no update by user is being attempted)
+	archivalEvent struct {
+		defaultBucket string
+		bucket        string
+		status        *shared.ArchivalStatus
 	}
 )
 
+// the following errors represents impossible code states that should never occur
 var (
-	errCannotProvideBucket       = &shared.BadRequestError{Message: "Request can only provide bucket name if request also gives status of enabled or disabled."}
-	errDisallowedBucketMetadata  = &shared.BadRequestError{Message: "Cannot set bucket owner or bucket retention (must update bucket manually)."}
-	errStateChangeToNeverEnabled = &shared.BadRequestError{Message: "Cannot change state to never enabled."}
-	errBucketNameUpdate          = &shared.BadRequestError{Message: "Cannot update bucket name after archival status has been moved out of never_enabled for the first time."}
-	errCannotProvideEmptyBucket  = &shared.BadRequestError{Message: "Cannot provide empty bucket name."}
-	errNoDefaultBucket           = &shared.BadRequestError{Message: "No default bucket provided, is cluster correctly configured for archival?"}
-	errCannotHandleStateChange   = &shared.BadRequestError{Message: "Encountered current state and update that cannot be handled (this should be impossible)"}
+	errInvalidState            = &shared.BadRequestError{Message: "Encountered illegal state: archival is enabled but bucket is not set (should be impossible)"}
+	errInvalidEvent            = &shared.BadRequestError{Message: "Encountered illegal event: default bucket is not set (should be impossible)"}
+	errCannotHandleStateChange = &shared.BadRequestError{Message: "Encountered current state and event that cannot be handled (should be impossible)"}
 )
 
-func neverEnabledState() *archivalConfigState {
-	return &archivalConfigState{
+// the following errors represents bad user input
+var (
+	errDisallowedBucketMetadata = &shared.BadRequestError{Message: "Cannot set bucket owner or bucket retention (must update bucket manually)"}
+	errBucketNameUpdate         = &shared.BadRequestError{Message: "Cannot update existing bucket name"}
+)
+
+func neverEnabledState() *archivalState {
+	return &archivalState{
 		bucket: "",
-		status: shared.ArchivalStatusNeverEnabled,
+		status: shared.ArchivalStatusDisabled,
 	}
 }
 
-func registerRequestToArchivalConfig(request *shared.RegisterDomainRequest, defaultBucket string) (*archivalConfigUpdate, error) {
-	requestArchivalConfig := &archivalConfigUpdate{
-		defaultBucket:   defaultBucket,
-		requestedBucket: request.ArchivalBucketName,
-		requestedStatus: request.ArchivalStatus,
+func registerToEvent(request *shared.RegisterDomainRequest, defaultBucket string) (*archivalEvent, error) {
```

Comment:
I like having a constructor like newArchivalRegisterEvent(param1, param2...) to this approach - (1) less coupling between this and the rpc handler (2) registerToEvent() as a top level function doesn't make a lot of sense for the reader

Or alternatively, rename this method to toArchivalRegisterEvent() and make this a class method inside your workflowHandler class.

Reply:
I picked the second option because logically I am thinking of this file as just an extension workflowHandler.go. Despite making these member methods on workflowHandler, I still left them in this file rather than moving them to workflowHandler.go file.